src/strategies/localStrategy.ts

```diff
@@ -15,27 +29,94 @@ import { SpidLevelEnum } from "../../generated/backend/SpidLevel";
  */
 export const localStrategy = (
   validUsernameList: ReadonlyArray<FiscalCode>,
-  validPassword: string
+  validPassword: string,
+  isLollipopEnabled: boolean,
+  lollipopApiClient: ReturnType<LollipopApiClient>
 ): passport.Strategy =>
   new Strategy({ passReqToCallback: true }, (req, username, password, done) => {
-    if (
-      FiscalCode.is(username) &&
-      validUsernameList.includes(username) &&
-      validPassword === password
-    ) {
-      // Fake test user for password based logins
-      const testUser: SpidUser = {
-        authnContextClassRef: SpidLevelEnum["https://www.spid.gov.it/SpidL2"],
-        dateOfBirth: "2000-06-02",
-        familyName: "Rossi",
-        fiscalNumber: username,
-        getAcsOriginalRequest: () => req,
-        getAssertionXml: () => "<xml></xml>",
-        getSamlResponseXml: () => "<xml></xml>",
-        issuer: Object.keys(SPID_IDP_IDENTIFIERS)[0] as Issuer,
-        name: "Mario",
-      };
-      return done(null, testUser);
-    }
-    return done(null, false);
+    pipe(
+      TE.of(
+        FiscalCode.is(username) &&
+          validUsernameList.includes(username) &&
+          validPassword === password
+      ),
+      TE.chain(
+        TE.fromPredicate(
+          (isValidAuth) => isValidAuth,
+          () => new Error("Invalid credentials")
+        )
+      ),
+      TE.chain(() =>
+        TE.tryCatch(
+          () => lollipopLoginHandler(isLollipopEnabled, lollipopApiClient)(req),
+          E.toError
+        )
+      ),
+      TE.chain(
+        flow(
+          TE.fromPredicate(
+            (_): _ is undefined | LollipopLoginParams =>
+              LollipopLoginParams.is(_) || _ === undefined,
+            () => new Error("Invalid lollipop parameters")
+          )
+        )
+      ),
+      TE.chainW(
+        flow(
+          TE.fromPredicate(LollipopLoginParams.is, identity),
+          TE.chainW((_) =>
+            pipe(
+              TE.tryCatch(
+                () => jose.calculateJwkThumbprint(_.jwk, _.algo),
+                E.toError
+              ),
+              TE.map(
+                (thumbprint) => `${_.algo}-${thumbprint}` as NonEmptyString
+              )
+            )
+          ),
+          TE.orElse((_) => TE.of("_aTestValueRequestId" as NonEmptyString))
+        )
+      ),
```

Comment:
may I suggest this edit just to improve readability? instead of having a type guard that verifies if the response of the handler is undefined, I put in this suggestion an Option monad to handle the undefined use case. if the handler response is instead different from `LollipopLoginParams` type the either introduced in the suggestion will handler the error correctly. what do you think? 
```suggestion
            TE.chain(() =>
        pipe(
          TE.tryCatch(
            () =>
              lollipopLoginHandler(isLollipopEnabled, lollipopApiClient)(req),
            E.toError
          ),
          TE.map(O.fromNullable)
        )
      ),
      TE.chainW(
        flow(
          O.map((handlerResponse) =>
            pipe(
              LollipopLoginParams.decode(handlerResponse),
              E.mapLeft(() => new Error("Invalid lollipop parameters"))
            )
          ),
          TE.of
        )
      ),
      TE.chainW(
        flow(
          O.fold(
            () => TE.of("_aTestValueRequestId" as NonEmptyString),
            flow(
              E.fold(
                (error) => TE.left(error),
                (lollipopLoginParams) =>
                  pipe(
                    TE.tryCatch(
                      () =>
                        jose.calculateJwkThumbprint(
                          lollipopLoginParams.jwk,
                          lollipopLoginParams.algo
                        ),
                      E.toError
                    ),
                    TE.map(
                      (thumbprint) =>
                        `${lollipopLoginParams.algo}-${thumbprint}` as NonEmptyString
                    )
                  )
              )
            )
          )
        )
      ),
```

Reply:
Nice but a little bit overwhelming. The suggested code respect the pure functional paradigm. But in this case create only a more complex algorithm, whit an indentation hell and a complex type composition `TE<E<O<data>>>` only to does't handle an undefined. We can instead semplify the solution using this:
``` typescript
TE.chain(
  flow(
    TE.fromPredicate(
      (_): _ is undefined | LollipopLoginParams =>
        LollipopLoginParams.is(_) || _ === undefined,
      () => new Error("Invalid lollipop parameters")
    )
  )
),
TE.chainW(
  flow(
    TE.of,
    TE.chainW((_) =>
      pipe(
        TE.tryCatch(
          async () =>
            (_ === undefined
              ? "_aTestValueRequestId"
              : `${_.algo}-${await jose.calculateJwkThumbprint(
                  _.jwk,
                  _.algo
                )}`) as NonEmptyString,
          E.toError
        )
      )
    )
  )
),
```
cc. @gquadrati

Reply:
fix in [75cac17](https://github.com/pagopa/io-backend/pull/1071/commits/75cac17d47f869bd3d8ab6bb70d58c503847be68)